src/icp_v2.cc

```diff
@@ -443,14 +441,20 @@ icpDenyAccess(Ip::Address &from, char *url, int reqnum, int fd)
 bool
 icpAccessAllowed(Ip::Address &from, HttpRequest * icp_request)
 {
-    /* absent any explicit rules, we deny all */
-    if (!Config.accessList.icp)
+    if (!Config.accessList.icp) {
+        debugs(12, 2, "Access Denied due to lack of ICP access rules.");
         return false;
+    }
 
     ACLFilledChecklist checklist(Config.accessList.icp, icp_request, nullptr);
     checklist.src_addr = from;
     checklist.my_addr.setNoAddr();
-    return checklist.fastCheck().allowed();
+    const auto &answer = checklist.fastCheck();
+    if (answer.allowed())
+        return true;
+
+    debugs(12, 2, "Access Denied for " << from << " by " << answer.lastCheckDescription() << ".");
```

Comment:
This small icpAccessAllowed() refactoring preserves icpDenyAccess()-related behavior, including debugging. We refactored so that we do not have to pass `answer` to icpDenyAccess(). Ideally, the two functions should be merged instead of relying on icpAccessAllowed() caller to call icpDenyAccess() after icpAccessAllowed() returns false, but doing that right requires a lot more changes, including caller changes that we want to avoid in this PR.